src/btf_parser.cpp

```diff
@@ -1081,3 +1085,393 @@ std::map<std::string, size_t> parse_btf_map_sections(const btf_type_data& btf_da
     }
     return map_offset_to_descriptor_index;
 }
+
+/**
+ * @brief Write a type into a vector and grow it.
+ *
+ * @tparam T Type to write
+ * @param[in] btf The .BTF section
+ * @param[in] value The value to write
+ */
+template <typename T>
+void _write_btf(std::vector<std::byte>& btf, const T& value) {
+    size_t length = 0;
+    size_t offset = btf.size();
+    if constexpr (std::is_same<T, std::string>::value) {
+        length = value.length();
+        btf.resize(offset + length + 1);
+        memcpy(btf.data() + offset, value.c_str(), length + 1);
+    } else if constexpr (std::is_same<T, std::vector<std::byte>>::value) {
+        length = value.size();
+        btf.resize(offset + length);
+        memcpy(btf.data() + offset, value.data(), length);
+    } else {
+        length = sizeof(T);
+        btf.resize(offset + length);
+        memcpy(btf.data() + offset, &value, length);
+    }
+}
+
+std::vector<std::byte> btf_serialize_types(const std::vector<btf_kind>& btf_kind) {
+    std::vector<std::byte> btf;
+    std::vector<std::byte> string_table_bytes;
+    std::map<std::string, uint32_t> string_table_map;
+    std::vector<std::byte> type_table_bytes;
+
+    auto string_to_offset = [&](const std::optional<std::string>& str) -> uint32_t {
+        if (!str) {
+            return 0;
+        }
+        auto it = string_table_map.find(*str);
+        if (it != string_table_map.end()) {
+            return it->second;
+        }
+        size_t offset = string_table_bytes.size();
+        _write_btf(string_table_bytes, *str);
+        string_table_map[*str] = static_cast<uint32_t>(offset);
+        return static_cast<uint32_t>(offset);
+    };
+
+    string_to_offset("");
+
+    auto pack_btf_int_data = [](bool is_signed, bool is_char, bool is_bool, size_t offset, size_t bits) {
+        uint32_t value = 0;
+        value |= is_signed ? BTF_INT_SIGNED : 0;
+        value |= is_char ? BTF_INT_CHAR : 0;
+        value |= is_bool ? BTF_INT_BOOL : 0;
+        value = value << 24;
+        value |= offset & UINT8_MAX << 16;
+        value |= bits & UINT8_MAX;
+        return value;
+    };
+
+    for (const auto& kind : btf_kind) {
+        auto pack_btf_info = [&](size_t vlen = 0, bool flag = false) {
+            union {
+                struct {
+                    int vlen : 16;
+                    int unused : 8;
+                    int kind : 5;
+                    int unused2 : 2;
+                    int flag : 1;
+                };
+                uint32_t value;
+            } info;
+            info.vlen = vlen;
+            info.kind = kind.index();
+            info.flag = flag;
+            return info.value;
+        };
```

Comment:
I think it's better to have an external function, and take `kind` as an argument.